### PR TITLE
Updated the meaning of an incomplete versionstamp to match existing bindings

### PR DESCRIPTION
**Pull request by dimitribouniol**
As is pointed out in https://forums.foundationdb.org/t/implementing-versionstamps-in-bindings/250, existing bindings use all `0xFF` rather than `0x00` for incomplete version stamps, likely because they will always come after any other version stamps lexicographically anyways.

### CODE REVIEW

**Review by kirilltitov — COMMENTED** [review]
Solid work, thank you! BTW, while we're here, could you please update readme.md? There is a section about FDB tuples with all possible tuple types, having versionstamp there would be perfect.

**Comment by dimitribouniol** [comment]
It's at this moment that I realize that getVersionStamp should either have a commit flag, or should return a tuple future with the transaction - not a big deal though 😅

That said, am I doing something wrong, or is there a better way of avoiding the casts between the sync version and the async version of each method? Here is a situation where you might write to a versionstamped key, then use that versionstamp for something in a follow-up transaction, but it feels a bit forced with the `as EventLoopFuture<...>`s everywhere...

```swift
        let versionstampedKeyValueFuture: EventLoopFuture<String?> = (fdb.withTransaction(on: self.eventLoop) { transaction in
            return try transaction
                .set(versionstampedKey: asyncKey, value: getBytes(asyncValue.utf8))
                .flatMap { transaction in
                    let versionstamp = transaction.getVersionstamp() as EventLoopFuture<FDB.Versionstamp>
                    return transaction.commit().flatMap { versionstamp }
                }
        } as EventLoopFuture<FDB.Versionstamp>)
            .flatMap { versionstamp in
                var versionstamp = versionstamp
                XCTAssertNil(versionstamp.userData)
                versionstamp.userData = 42

                let updatedKey = subspace[versionstamp]["myAsynckey"]
                
                return fdb.withTransaction(on: self.eventLoop) { transaction in
                    return transaction
                        .get(key: updatedKey, commit: true)
                        .map { maybeBytes, transaction in
                            guard let bytes = maybeBytes else {
                                return nil
                            }
                            return try? bytes.cast() as String
                        }
                }
        }
```

**Comment by kirilltitov** [comment]
Yes, I also like the idea of adding commit flag to that method (though it should be additive change, hence new method), however, I don't know if that argument should have a default value (I don't think so).

Also, I've just realised that you've marked `set(versionstampedKey:value:commit:) -> EventLoopFuture<AnyFDBTransaction>` as throwing, which isn't reasonable in event loop context. Let's make it non-throwable, but instead fail the resulting future.

**Comment by dimitribouniol** [comment]
My lack of experience using futures is showing 😛 I'll make the changes right now.

**Comment by kirilltitov** [comment]
It's fine :) BTW I wrote a big and comprehensive article on futures / NIO ideology: https://kirilltitov.com/en/blog/2019/futures

**Comment by dimitribouniol** [comment]
Thanks for that, I will be reading that more times than once 😅

In other news, I made the improvements, it's looking quite a bit better... (Also fixed an issue where synchronous getVersionstamp() was not actually synchronous...)

**Comment by kirilltitov** [comment]
Fun fact: authorisation and comments services on my website are written in Swift, and they both use FoundationDB as database :)

File: Sources/FDB/Transaction/Transaction+NIO.swift
Comment (kirilltitov) [inline Sources/FDB/Transaction/Transaction+NIO.swift:417]: Do you actually need a `switch` for a simple comparison? :D I would do something like this or simple `if/else`
```suggestion
        return shouldCommit
            ? self.commit().flatMap { promise.futureResult }
            : promise.futureResult
```

File: Tests/FDBTests/FDBTests.swift
Comment (kirilltitov) [inline Tests/FDBTests/FDBTests.swift:142]: You don't really need this `transaction` argument at 141, as it's the same instance as on line 138. `set` methods are designed to return future to transaction for cases when user wants to handle the transaction manually, without `withTransaction`.

File: Tests/FDBTests/FDBTests.swift
Comment (kirilltitov) [inline Tests/FDBTests/FDBTests.swift:144]: This looks very ugly :D Try to avoid this casting. It can be done, for example, by specifying the return type at 138 and 141.

File: Tests/FDBTests/FDBTests.swift
Comment (kirilltitov) [inline Tests/FDBTests/FDBTests.swift:168]: Two empty lines

**Review by kirilltitov — APPROVED** [review]
LG%C
Cool. Just a couple of concerns/questions/refinements.
Also I would like to ask you to squash commits before merging.

File: Tests/FDBTests/FDBTests.swift
Comment (kirilltitov) [reply Tests/FDBTests/FDBTests.swift:144]: I've done it this way (code from my sandbox):

```swift
let result: String? = try connection
    .withTransaction(on: eventLoop) { (transaction: AnyFDBTransaction) in
        try! transaction
            .set(versionstampedKey: asyncKey, value: Bytes(asyncValue.utf8))
            .flatMap { (_) -> EventLoopFuture<FDB.Versionstamp> in
                let versionstamp: EventLoopFuture<FDB.Versionstamp> = transaction.getVersionstamp()
                return transaction.commit().flatMap { versionstamp }
            }
    }
    .flatMap { (versionstamp: FDB.Versionstamp) -> EventLoopFuture<String?> in
        var versionstamp = versionstamp
        dump(versionstamp)
        versionstamp.userData = 42
        dump(versionstamp)

        return connection.withTransaction(on: eventLoop) { transaction in
            transaction
                .get(key: subspace[versionstamp]["myAsynckey"], commit: true)
                .map { (maybeBytes, _) -> String? in
                    guard let bytes = maybeBytes else {
                        return nil
                    }
                    return String(bytes: bytes, encoding: .ascii)!
                }
        }
    }
    .wait()
```

**Comment by dimitribouniol** [comment]
Squash every commit in this branch into a single one, or squash the fixes into their respective commits? I'm fine with either.

**Comment by kirilltitov** [comment]
I like single commit.

**Comment by dimitribouniol** [comment]
Single commit it is!

**Comment by dimitribouniol** [comment]
I re-wrote the tests so they are better organized and more exhaustive (testing sync/async, commit/no-commit, invalid keys, and multiple versionstamp writes in a single transaction).

Also, I removed the `commit:` option from the synchronous `getVersionstamp()`, since it would be impossible to not commit it at this point (since it will wait indefinitely for the transaction to be committed)

**Comment by dimitribouniol** [comment]
(typos in the readme)

File: Sources/FDB/Future/Future+Bytes.swift
Comment (dimitribouniol) [inline Sources/FDB/Future/Future+Bytes.swift:51]: Bringing this to your attention — I'm assuming this is not allowed re: SemVer, but I accidentally introduced this as public in #59, failing to notice the "value" version was not.

File: Sources/FDB/AnyFDBTransaction.swift
Comment (dimitribouniol) [inline Sources/FDB/AnyFDBTransaction.swift:306]: Should this return the transaction as well? I noticed other methods do this…
```suggestion
    func getVersionstamp(commit shouldCommit: Bool) -> EventLoopFuture<(FDB.Versionstamp, AnyFDBTransaction)>
```

What about `func getVersionstamp() -> EventLoopFuture<FDB.Versionstamp>`?

File: Sources/FDB/Future/Future+Bytes.swift
Comment (kirilltitov) [reply Sources/FDB/Future/Future+Bytes.swift:51]: It doesn't matter, `FDB.Future` class is internal already, and also in modules `public` should be stated explicitly, otherwise it's `internal` (might be mistaken, but from my experience this is how it is).

File: Sources/FDB/AnyFDBTransaction.swift
Comment (kirilltitov) [reply Sources/FDB/AnyFDBTransaction.swift:306]: Oof. Tough question. I gotta think about it.

File: Sources/FDB/Future/Future+Bytes.swift
Comment (dimitribouniol) [reply Sources/FDB/Future/Future+Bytes.swift:51]: I'll leave it with `internal` for consistency then.

File: Sources/FDB/AnyFDBTransaction.swift
Comment (dimitribouniol) [reply Sources/FDB/AnyFDBTransaction.swift:306]: On the plus side, I'm likely the only user of Versionstamps right now, so probably no one will notice if the return signature changes a patch release in 🙃

Comment (kirilltitov) [reply Sources/FDB/AnyFDBTransaction.swift:306]: Yes, as for this feature, semver stability doesn't really bother me right now :) Yes, let's return a tuple of a versionstamp and a transaction, for the sake of consistency.

Comment (dimitribouniol) [reply Sources/FDB/AnyFDBTransaction.swift:306]: One more data point: `get` has both signatures: 
image

Comment (dimitribouniol) [reply Sources/FDB/AnyFDBTransaction.swift:306]: Actually, it can be solved by adding the following to the extension on AnyFDBTransaction:
```swift
    /// Returns a future for the versionstamp which was used by any versionstamp operations
    /// in this transaction. Unlike the synchronous version of the same name, this method
    /// does _not_ commit by default.
    ///
    /// - returns: EventLoopFuture with future FDB.Versionstamp value
    func getVersionstamp() -> EventLoopFuture<(FDB.Versionstamp, AnyFDBTransaction)> {
        return self
            .getVersionstamp()
            .map { ($0, self) }
    }

    /// Returns a future for the versionstamp which was used by any versionstamp operations
    /// in this transaction, and optionally commit the transaction right afterwards
    ///
    /// - returns: EventLoopFuture with future FDB.Versionstamp value
    func getVersionstamp(commit shouldCommit: Bool) -> EventLoopFuture<(FDB.Versionstamp, AnyFDBTransaction)> {
        return self
            .getVersionstamp(commit: shouldCommit)
            .map { ($0, self) }
    }
```

Comment (kirilltitov) [reply Sources/FDB/AnyFDBTransaction.swift:306]: Wellllll, I'm afraid, it was done out of semver desperation or smth, I've had much more pain from these two definitions than convenience.

From the other side... Some use transactions directly, some with wrapper... Dammit. I gotta sleep with this dilemma. Don't merge pls, I haven't decided yet.

Comment (dimitribouniol) [reply Sources/FDB/AnyFDBTransaction.swift:306]: I don't have a merge button, so you are in luck 😉

With the snippet I posted above, the user has both variations at their disposal, and it seems like the type system is smart enough to differentiate between then, seeing as I didn't need to touch the tests...

(though they are casting to begin with since it was necessary to get the rest of the type system happy anyways:
```swift
                    .flatMap { _ -> EventLoopFuture<FDB.Versionstamp> in
                        return transaction.getVersionstamp(commit: true)
                    }
```

Comment (kirilltitov) [reply Sources/FDB/AnyFDBTransaction.swift:306]: That's weird, I approved your PR, so you should be able to merge it. Funny.
Ye, I like two versions with different arguments, let's do that.

Comment (dimitribouniol) [reply Sources/FDB/AnyFDBTransaction.swift:306]: (I believe it's because I don't have write access, which I absolutely don't need 😉)

Changes have been pushed!

**Comment by dimitribouniol** [comment]
Thanks!

**Comment by kirilltitov** [comment]
No u :D

**Pull request merged**
